### PR TITLE
add class mapping for multi class segmentation

### PR DESCRIPTION
**Pull request by chaymaaBOUSNAH**

### CODE REVIEW

**Comment by Mattwmaster58** [comment]
This doesn't seem to be working for me, at least using 3 channel pngs as masks

**Comment by chaymaaBOUSNAH** [comment]
> This doesn't seem to be working for me, at least using 3 channel pngs as masks

how many classes do you have? you must ensure that you have mapped all classes, can you share the mapping dict you have set?

**Comment by Mattwmaster58** [comment]
Upon closer investigation, the issue appears to be related to my mask image format as its in palette mode already. As I understand it, no further conversion should be necessary then (ie, no need to use mask_to_class). Although even after converting with convert("RGB"), other things break that I'm not sure are related. Here's one mask, I'm not sure that GH won't re-encode.


Here's the code I added in `__getitem__`:
```
        mask_mode = mask.mode

        assert img.size == mask.size, \
            f'Image and mask {name} should be the same size, but are {img.size} and {mask.size}'

        img = self.preprocess(img, self.scale, is_mask=False)
        mask = self.preprocess(mask, self.scale, is_mask=True)

        if mask_mode != "P":  # if pallete mode, no need to convert a RGB color because its already correct at this point
            # mapping the class colors
            mask = self.mask_to_class(mask, self.mapping)
```

**Comment by chaymaaBOUSNAH** [comment]
> I don't think PIL is handling the conversion to RGB correctly, the mask np.array has absurdly large values in it

Can you check your mask shape and values after conversion to RGB? Can you share the mapping dict you set for loading dataset in train.py ?

**Comment by sevimcengiz** [comment]
how do we need to arrange for mapping values if we have more than 3 classess?

**Comment by chaymaaBOUSNAH** [comment]
> how do we need to arrange for mapping values if we have more than 3 classess?

You must map each class color code to an index, for example: if you have 5 classes with the following colors: Black, blue, red, aqua, and magenta. the mapping dict must be: mapping = {(0, 0, 0): 0, (0, 0, 255): 1, (255, 0, 0): 2, (0, 255, 255): 3, (255, 0, 255): 4}

**Comment by chaymaaBOUSNAH** [comment]
I didn't test the Carnava dataset, but if you are talking about the mask shape after mapping, it must be (h, w) and there will be no channel dim.
 we do the transpose to have the shape (channels, height, width) instead of (height, width, channels) because most Pytorch functions must have this input format, like the Cross-Entropy loss ...
I removed /255 because mapping do already the job (after mapping masks contain only indexes of each class color).
I hope my answer is clear!

**Comment by onway666** [comment]
> how do we need to arrange for mapping values if we have more than 3 classess?



> > how do we need to arrange for mapping values if we have more than 3 classess?
> 
> You must map each class color code to an index, for example: if you have 5 classes with the following colors: Black, blue, red, aqua, and magenta. the mapping dict must be: mapping = {(0, 0, 0): 0, (0, 0, 255): 1, (255, 0, 0): 2, (0, 255, 255): 3, (255, 0, 255): 4}

Doesn't this operation create some kind of association between different classes? What if my classes have no relationship with each other?

**Comment by onway666** [comment]
What should I do if my original mask is a four-channel image where each channel is a binary image?

**Comment by BenjaminSchaar** [comment]
I am not 100 sure how to use more than 2 classes:
Should I do more than changing the parameter classes ?
